pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/SegmentPushUtils.java

```diff
@@ -0,0 +1,349 @@
+          retryWaitMs = spec.getPushJobSpec().getPushRetryIntervalMillis();
+        }
+        RetryPolicies.exponentialBackoffRetryPolicy(attempts, retryWaitMs, 5).attempt(() -> {
+          try {
+            SimpleHttpResponse response = FILE_UPLOAD_DOWNLOAD_CLIENT
+                .sendSegmentUri(FileUploadDownloadClient.getUploadSegmentURI(controllerURI), segmentUri,
+                    FileUploadDownloadClient.makeAuthHeader(spec.getAuthToken()),
+                    FileUploadDownloadClient.makeTableParam(tableName),
+                    FileUploadDownloadClient.DEFAULT_SOCKET_TIMEOUT_MS);
+            LOGGER.info("Response for pushing table {} segment uri {} to location {} - {}: {}", tableName, segmentUri,
+                controllerURI, response.getStatusCode(), response.getResponse());
+            return true;
+          } catch (HttpErrorStatusException e) {
+            int statusCode = e.getStatusCode();
+            if (statusCode >= 500) {
+              // Temporary exception
+              LOGGER.warn("Caught temporary exception while pushing table: {} segment uri: {} to {}, will retry",
+                  tableName, segmentUri, controllerURI, e);
+              return false;
+            } else {
+              // Permanent exception
+              LOGGER.error("Caught permanent exception while pushing table: {} segment uri: {} to {}, won't retry",
+                  tableName, segmentUri, controllerURI, e);
+              throw e;
+            }
+          } finally {
+            if (spec.isCleanUpOutputDir()) {
+              outputDirFS.delete(segmentURI, true);
+            }
+          }
+        });
+      }
+    }
+  }
+
+  /**
+   * This method takes a map of segment downloadURI to corresponding tar file path, and push those segments in
+   * metadata mode.
+   * The steps are:
+   * 1. Download segment from tar file path;
+   * 2. Untar segment metadata and creation meta files from the tar file to a segment metadata directory;
+   * 3. Tar this segment metadata directory into a tar file
+   * 4. Generate a POST request with segmentDownloadURI in header to push tar file to Pinot controller.
+   *
+   * @param spec is the segment generation job spec
+   * @param fileSystem is the PinotFs used to copy segment tar file
+   * @param segmentUriToTarPathMap contains the map of segment DownloadURI to segment tar file path
+   * @throws Exception
+   */
+  public static void sendSegmentUriAndMetadata(SegmentGenerationJobSpec spec, PinotFS fileSystem,
+      Map<String, String> segmentUriToTarPathMap)
+      throws Exception {
+    String tableName = spec.getTableSpec().getTableName();
+    LOGGER.info("Start pushing segment metadata: {} to locations: {} for table {}", segmentUriToTarPathMap,
+        Arrays.toString(spec.getPinotClusterSpecs()), tableName);
+    for (String segmentUriPath : segmentUriToTarPathMap.keySet()) {
+      String tarFilePath = segmentUriToTarPathMap.get(segmentUriPath);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.